app/services/user_session_context.rb

```diff
@@ -7,12 +7,12 @@ def self.authentication_context?(context)
     context == AUTHENTICATION_CONTEXT
   end
 
-  def self.reauthentication_context?(context)
-    context == REAUTHENTICATION_CONTEXT
+  def self.reauthentication_context?(context, user_fully_authenticated)
+    context == REAUTHENTICATION_CONTEXT && user_fully_authenticated
```

Comment:
Maybe I'm missing something obvious, but what scenarios are we preventing by checking `user_fully_authenticated` here?

Reply:
Put another way: I get that we don't expect reauthentication except for a fully-authenticated user, but I'm thinking we should have that check happen _before_ we set the reauthentication context?

Reply:
This part I'm less sure about. My intention was to be a bit defensive and I also considered raising an exception if the session context is `reauthentication` and the user is not fully authenticated. There _shouldn't_ be a case where that happens and I don't think it can happen with our current usage, so it's almost entirely a guard against future changes.

The authorization aspect of the session context makes me a bit nervous and adding a guardrail here makes me feel a bit better, but I'm not sure if it is worth doing in this manner.

Reply:
It seems like the two places we're setting `user_session[:context] = 'reauthentication'` both have guards for checking the user is fully authenticated?

1. Assigned: https://github.com/18F/identity-idp/blob/ba0bdd0d82501b184e467f3edfc4a19c1d252c51/app/controllers/accounts_controller.rb#L27
   1. Guarded: https://github.com/18F/identity-idp/blob/ba0bdd0d82501b184e467f3edfc4a19c1d252c51/app/controllers/accounts_controller.rb#L3
2. Assigned: https://github.com/18F/identity-idp/blob/3fc7f043d311173ab82e81bf8c0cacb1e4366cf3/app/controllers/concerns/reauthentication_required_concern.rb#L31
   1. Guarded: https://github.com/18F/identity-idp/blob/3fc7f043d311173ab82e81bf8c0cacb1e4366cf3/app/controllers/concerns/reauthentication_required_concern.rb#L6

Reply:
Reverted out all these changes in a753fda